File: apps/demo9_softBody/generate_softbody.mjs
#!/usr/bin/env node
/**
 * Generate a pre-baked soft-body asset file for demo9's PBD plugin.
 *
 * The asset contains all structural data that used to be generated at runtime:
 *   - particle rest positions (axis-aligned, centered at origin, no rotation)
 *   - surface triangle indices (cube shell)
 *   - shape-matching cluster data (sparse centers, Chebyshev membership)
 *
 * Cluster center selection: greedy, minimum Chebyshev spacing `clusterSpacing`
 * between centers. Each center's cluster = all particles within Chebyshev
 * radius `clusterRadius`. With clusterSpacing=clusterRadius, every particle is
 * covered by at least one cluster (full coverage, minimal overlap). With
 * clusterSpacing=0, every particle is a center (densest, maximum overlap).
 *
 * Usage:
 *   node generate_softbody.mjs [--gridN 5] [--cellSize 0.4]
 *                              [--clusterRadius 2] [--clusterSpacing 2]
 *                              [--mass 1.0] [--output softbody_asset.json]
 *
 * A Python equivalent (generate_softbody.py) is also provided.
 */

import { writeFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { parseArgs } from 'node:util';

function idxAt(i, j, k, n) {
    return i + j * n + k * n * n;
}

function buildSurfaceIndices(n) {
    const out = [];
    const at = (i, j, k) => idxAt(i, j, k, n);

    for (let j = 0; j < n - 1; j++) {
        for (let k = 0; k < n - 1; k++) {
            const a = at(0, j, k), b = at(0, j + 1, k), c = at(0, j + 1, k + 1), d = at(0, j, k + 1);
            out.push(a, b, c, a, c, d);
            const a2 = at(n - 1, j, k), b2 = at(n - 1, j + 1, k), c2 = at(n - 1, j + 1, k + 1), d2 = at(n - 1, j, k + 1);
            out.push(a2, c2, b2, a2, d2, c2);
        }
    }
    for (let i = 0; i < n - 1; i++) {
        for (let k = 0; k < n - 1; k++) {
            const a = at(i, 0, k), b = at(i + 1, 0, k), c = at(i + 1, 0, k + 1), d = at(i, 0, k + 1);
            out.push(a, c, b, a, d, c);
            const a2 = at(i, n - 1, k), b2 = at(i + 1, n - 1, k), c2 = at(i + 1, n - 1, k + 1), d2 = at(i, n - 1, k + 1);
            out.push(a2, b2, c2, a2, c2, d2);
        }
    }
    for (let i = 0; i < n - 1; i++) {
        for (let j = 0; j < n - 1; j++) {
            const a = at(i, j, 0), b = at(i + 1, j, 0), c = at(i + 1, j + 1, 0), d = at(i, j + 1, 0);
            out.push(a, b, c, a, c, d);
            const a2 = at(i, j, n - 1), b2 = at(i + 1, j, n - 1), c2 = at(i + 1, j + 1, n - 1), d2 = at(i, j + 1, n - 1);
            out.push(a2, c2, b2, a2, d2, c2);
        }
    }
    return out;
}

/**
 * Seeded LCG random number generator (deterministic — same seed → same output).
 * Returns a function producing floats in [0, 1).
 */
function makeRng(seed) {
    let state = (seed >>> 0) || 1;
    return () => {
        state = (Math.imul(state, 1664525) + 1013904223) >>> 0;
        return state / 0x100000000;
    };
}

function randInt(rng, min, max) {
    return Math.floor(rng() * (max - min + 1)) + min;
}

function shuffle(rng, arr) {
    for (let i = arr.length - 1; i > 0; i--) {
        const j = Math.floor(rng() * (i + 1));
        const tmp = arr[i]; arr[i] = arr[j]; arr[j] = tmp;
    }
    return arr;
}

/**
 * Option C: jittered sublattice + greedy dedup + forced coverage.
 *
 * 1. Generate candidate centers on a regular sublattice (spacing = spacingCells),
 *    each displaced by ±jitter cells (clamped to grid).
 * 2. Shuffle candidates (seeded) and greedily select, skipping any within
 *    Chebyshev distance `spacingCells` of an already-selected center.
 * 3. Coverage check: every particle must be within `radiusCells` of some center.
 *    Uncovered particles are force-added as new centers (may violate spacing —
 *    coverage is mandatory, spacing is a soft constraint).
 *
 * With radiusCells=R and spacingCells=S, adjacent centers share (2R+1−S) layers
 * of particles. For 1–2 layers shared: S = 2R or S = 2R−1.
 */
function selectCenters(gridN, radiusCells, spacingCells, jitter, rng) {
    // 1. Sublattice candidates with jitter
    const candidates = [];
    for (let k = 0; k < gridN; k += spacingCells) {
        for (let j = 0; j < gridN; j += spacingCells) {
            for (let i = 0; i < gridN; i += spacingCells) {
                const ci = Math.max(0, Math.min(gridN - 1, i + randInt(rng, -jitter, jitter)));
                const cj = Math.max(0, Math.min(gridN - 1, j + randInt(rng, -jitter, jitter)));
                const ck = Math.max(0, Math.min(gridN - 1, k + randInt(rng, -jitter, jitter)));
                candidates.push([ci, cj, ck]);
            }
        }
    }
    // Ensure the far edge is represented even if gridN isn't a multiple of spacing.
    if ((gridN - 1) % spacingCells !== 0) {
        for (let j = 0; j < gridN; j += spacingCells) {
            for (let i = 0; i < gridN; i += spacingCells) {
                candidates.push([
                    Math.max(0, Math.min(gridN - 1, i + randInt(rng, -jitter, jitter))),
                    Math.max(0, Math.min(gridN - 1, j + randInt(rng, -jitter, jitter))),
                    gridN - 1,
                ]);
            }
        }
        for (let i = 0; i < gridN; i += spacingCells) {
            for (let k = 0; k < gridN; k += spacingCells) {
                candidates.push([
                    Math.max(0, Math.min(gridN - 1, i + randInt(rng, -jitter, jitter))),
                    gridN - 1,
                    Math.max(0, Math.min(gridN - 1, k + randInt(rng, -jitter, jitter))),
                ]);
            }
        }
        for (let j = 0; j < gridN; j += spacingCells) {
            for (let k = 0; k < gridN; k += spacingCells) {
                candidates.push([
                    gridN - 1,
                    Math.max(0, Math.min(gridN - 1, j + randInt(rng, -jitter, jitter))),
                    Math.max(0, Math.min(gridN - 1, k + randInt(rng, -jitter, jitter))),
                ]);
            }
        }
    }

    // 2. Shuffle + greedy select with min spacing
    shuffle(rng, candidates);
    const centers = [];
    for (const [ci, cj, ck] of candidates) {
        let tooClose = false;
        for (const [si, sj, sk] of centers) {
            if (Math.abs(ci - si) <= spacingCells &&
                Math.abs(cj - sj) <= spacingCells &&
                Math.abs(ck - sk) <= spacingCells) {
                tooClose = true;
                break;
            }
        }
        if (!tooClose) centers.push([ci, cj, ck]);
    }

    // 3. Coverage check: force-add centers for uncovered particles
    let added = 0;
    for (let k = 0; k < gridN; k++) {
        for (let j = 0; j < gridN; j++) {
            for (let i = 0; i < gridN; i++) {
                let covered = false;
                for (const [ci, cj, ck2] of centers) {
                    if (Math.abs(i - ci) <= radiusCells &&
                        Math.abs(j - cj) <= radiusCells &&
                        Math.abs(k - ck2) <= radiusCells) {
                        covered = true;
                        break;
                    }
                }
                if (!covered) {
                    centers.push([i, j, k]);
                    added++;
                }
            }
        }
    }
    if (added > 0) {
        console.warn(`info: coverage check added ${added} extra center(s) for uncovered particles`);
    }

    return centers;
}

function generate(gridN, cellSize, clusterRadius, clusterSpacing, jitter, seed, mass) {
    const particleCount = gridN ** 3;
    const invMass = particleCount / mass;
    const half = (gridN - 1) / 2;

    // ── particle rest positions (axis-aligned, centered at origin) ───────
    const positions = [];
    for (let k = 0; k < gridN; k++) {
        for (let j = 0; j < gridN; j++) {
            for (let i = 0; i < gridN; i++) {
                positions.push([
                    (i - half) * cellSize,
                    (j - half) * cellSize,
                    (k - half) * cellSize,
                    invMass,
                ]);
            }
        }
    }

    // ── surface indices ─────────────────────────────────────────────────
    const surfaceIndices = buildSurfaceIndices(gridN);

    // ── select cluster centers (jittered sublattice + greedy + coverage) ─
    const rng = makeRng(seed);
    const centers = selectCenters(gridN, clusterRadius, clusterSpacing, jitter, rng);
    const clusterCount = centers.length;

    // First pass: count members per cluster
    const clusterSizes = new Uint32Array(clusterCount);
    for (let c = 0; c < clusterCount; c++) {
        const [ci, cj, ck] = centers[c];
        let cnt = 0;
        for (let di = -clusterRadius; di <= clusterRadius; di++) {
            const ni = ci + di;
            if (ni < 0 || ni >= gridN) continue;
            for (let dj = -clusterRadius; dj <= clusterRadius; dj++) {
                const nj = cj + dj;
                if (nj < 0 || nj >= gridN) continue;
                for (let dk = -clusterRadius; dk <= clusterRadius; dk++) {
                    const nk = ck + dk;
                    if (nk < 0 || nk >= gridN) continue;
                    cnt++;
                }
            }
        }
        clusterSizes[c] = cnt;
    }

    // Prefix-sum for offsets
    const offsets = new Uint32Array(clusterCount);
    let total = 0;
    for (let c = 0; c < clusterCount; c++) {
        offsets[c] = total;
        total += clusterSizes[c];
    }
    const clusterEntries = total;

    // Second pass: write cluster data + indices + rest offsets
    const clusters = [];
    const clusterIndices = new Uint32Array(clusterEntries);
    const restOffsets = [];

    for (let c = 0; c < clusterCount; c++) {
        const [ci, cj, ck] = centers[c];
        const cnt = clusterSizes[c];
        const offset = offsets[c];

        // Rest center of mass
        let comX = 0, comY = 0, comZ = 0;
        for (let di = -clusterRadius; di <= clusterRadius; di++) {
            const ni = ci + di;
            if (ni < 0 || ni >= gridN) continue;
            for (let dj = -clusterRadius; dj <= clusterRadius; dj++) {
                const nj = cj + dj;
                if (nj < 0 || nj >= gridN) continue;
                for (let dk = -clusterRadius; dk <= clusterRadius; dk++) {
                    const nk = ck + dk;
                    if (nk < 0 || nk >= gridN) continue;
                    const pid = idxAt(ni, nj, nk, gridN);
                    comX += positions[pid][0];
                    comY += positions[pid][1];
                    comZ += positions[pid][2];
                }
            }
        }
        comX /= cnt;
        comY /= cnt;
        comZ /= cnt;

        clusters.push({ restCom: [comX, comY, comZ], count: cnt, offset });

        let localIdx = 0;
        for (let di = -clusterRadius; di <= clusterRadius; di++) {
            const ni = ci + di;
            if (ni < 0 || ni >= gridN) continue;
            for (let dj = -clusterRadius; dj <= clusterRadius; dj++) {
                const nj = cj + dj;
                if (nj < 0 || nj >= gridN) continue;
                for (let dk = -clusterRadius; dk <= clusterRadius; dk++) {
                    const nk = ck + dk;
                    if (nk < 0 || nk >= gridN) continue;
                    const pid = idxAt(ni, nj, nk, gridN);
                    const slot = offset + localIdx;
                    clusterIndices[slot] = pid;
                    restOffsets[slot] = [
                        positions[pid][0] - comX,
                        positions[pid][1] - comY,
                        positions[pid][2] - comZ,
                    ];
                    localIdx++;
                }
            }
        }
    }

    return {
        meta: {
            gridN, cellSize, mass, seed, jitter,
            clusterRadius: clusterRadius * cellSize,
            clusterSpacing: clusterSpacing * cellSize,
            clusterRadiusCells: clusterRadius,
            clusterSpacingCells: clusterSpacing,
            particleCount, clusterCount, clusterEntries,
            surfaceVertexCount: surfaceIndices.length,
        },
        positions,
        surfaceIndices,
        clusters,
        clusterIndices: Array.from(clusterIndices),
        restOffsets,
    };
}

const { values } = parseArgs({
    options: {
        gridN:          { type: 'string', default: '5' },
        cellSize:       { type: 'string', default: '0.4' },
        clusterRadius:  { type: 'string', default: '0.8' },   // meters
        clusterSpacing: { type: 'string', default: '' },      // default: = clusterRadius
        jitter:         { type: 'string', default: '1' },     // cells
        seed:           { type: 'string', default: '42' },    // RNG seed
        mass:           { type: 'string', default: '1.0' },
        output:         { type: 'string', default: 'softbody_asset.json' },
    },
});

const gridN = parseInt(values.gridN, 10);
const cellSize = parseFloat(values.cellSize);
const clusterRadiusM = parseFloat(values.clusterRadius);
const clusterSpacingM = values.clusterSpacing ? parseFloat(values.clusterSpacing) : clusterRadiusM;
const jitter = parseInt(values.jitter, 10);
const seed = parseInt(values.seed, 10);
const mass = parseFloat(values.mass);

// Convert meter values to grid cells (quantized — actual physical size may differ).
const clusterRadius = Math.max(1, Math.round(clusterRadiusM / cellSize));
const clusterSpacing = Math.max(1, Math.round(clusterSpacingM / cellSize));

if (gridN < 2) { console.error(`error: --gridN must be >= 2, got ${gridN}`); process.exit(1); }
if (cellSize <= 0) { console.error(`error: --cellSize must be > 0, got ${cellSize}`); process.exit(1); }
if (clusterRadiusM <= 0) { console.error(`error: --clusterRadius must be > 0, got ${clusterRadiusM}`); process.exit(1); }
if (clusterRadius > gridN - 1) { console.error(`error: --clusterRadius ${clusterRadiusM}m = ${clusterRadius} cells, must be < gridN (${gridN}) = ${(gridN-1)*cellSize}m`); process.exit(1); }
if (clusterSpacingM < 0) { console.error(`error: --clusterSpacing must be >= 0, got ${clusterSpacingM}`); process.exit(1); }
if (jitter < 0) { console.error(`error: --jitter must be >= 0, got ${jitter}`); process.exit(1); }

// Warn if quantization changed the actual value.
if (clusterRadius * cellSize !== clusterRadiusM) {
    console.warn(`warning: --clusterRadius ${clusterRadiusM}m quantized to ${clusterRadius} cells = ${(clusterRadius * cellSize).toFixed(3)}m`);
}
if (clusterSpacing * cellSize !== clusterSpacingM) {
    console.warn(`warning: --clusterSpacing ${clusterSpacingM}m quantized to ${clusterSpacing} cells = ${(clusterSpacing * cellSize).toFixed(3)}m`);
}

const data = generate(gridN, cellSize, clusterRadius, clusterSpacing, jitter, seed, mass);
const outPath = resolve(values.output);
writeFileSync(outPath, JSON.stringify(data, null, 2), 'utf-8');

const m = data.meta;
const cubeEdge = (m.gridN - 1) * m.cellSize;
const maxCluster = (2 * m.clusterRadiusCells + 1) ** 3;
const sharedLayers = Math.max(0, 2 * m.clusterRadiusCells + 1 - m.clusterSpacingCells);
console.log(
    `Generated ${values.output}:\n` +
    `  gridN=${m.gridN}  cellSize=${m.cellSize}m  cubeEdge=${cubeEdge.toFixed(3)}m  cubeVolume=${(cubeEdge ** 3).toFixed(3)}m³\n` +
    `  clusterRadius=${m.clusterRadius}m (${m.clusterRadiusCells} cells)  clusterSpacing=${m.clusterSpacing}m (${m.clusterSpacingCells} cells)` +
    `  jitter=${m.jitter}  seed=${m.seed}  mass=${m.mass}\n` +
    `  particles=${m.particleCount}  clusters=${m.clusterCount}  clusterEntries=${m.clusterEntries}\n` +
    `  maxClusterMembers=${maxCluster}  sharedLayers≈${sharedLayers}  surfaceTriangles=${m.surfaceVertexCount / 3}\n` +
    `  perParticleMass=${(m.mass / m.particleCount).toFixed(6)} kg  invMass=${(m.particleCount / m.mass).toFixed(1)}`
);

File: apps/demo9_softBody/generate_softbody.py
#!/usr/bin/env python3
"""Generate a pre-baked soft-body asset file for demo9's PBD plugin.

The asset contains all structural data that used to be generated at runtime:
  - particle rest positions (axis-aligned, centered at origin, no rotation)
  - surface triangle indices (cube shell)
  - shape-matching cluster data (sparse centers, Chebyshev membership)

Cluster center selection: greedy, minimum Chebyshev spacing `clusterSpacing`
between centers. Each center's cluster = all particles within Chebyshev
radius `clusterRadius`. With clusterSpacing=clusterRadius, every particle is
covered by at least one cluster (full coverage, minimal overlap). With
clusterSpacing=0, every particle is a center (densest, maximum overlap).

The runtime PbdManager loads this file, applies a random rotation + the
entity's Transform translation, and writes the result to GPU buffers.
Changing gridN / cellSize / clusterRadius / clusterSpacing requires re-running
this script — they are not runtime-tweakable component fields.

Usage:
    python generate_softbody.py [--gridN 5] [--cellSize 0.4]
                                [--clusterRadius 2] [--clusterSpacing 2]
                                [--mass 1.0] [--output softbody_asset.json]
"""

import argparse
import json
import sys
from typing import List, Dict, Any, Tuple, Callable


def idx_at(i: int, j: int, k: int, n: int) -> int:
    return i + j * n + k * n * n


def build_surface_indices(n: int) -> List[int]:
    """Cube shell triangulation — 6 faces, 2 triangles per cell, consistent winding."""
    out: List[int] = []
    at = lambda i, j, k: idx_at(i, j, k, n)

    for j in range(n - 1):
        for k in range(n - 1):
            a, b, c, d = at(0, j, k), at(0, j+1, k), at(0, j+1, k+1), at(0, j, k+1)
            out.extend([a, b, c, a, c, d])
            a2, b2, c2, d2 = at(n-1, j, k), at(n-1, j+1, k), at(n-1, j+1, k+1), at(n-1, j, k+1)
            out.extend([a2, c2, b2, a2, d2, c2])

    for i in range(n - 1):
        for k in range(n - 1):
            a, b, c, d = at(i, 0, k), at(i+1, 0, k), at(i+1, 0, k+1), at(i, 0, k+1)
            out.extend([a, c, b, a, d, c])
            a2, b2, c2, d2 = at(i, n-1, k), at(i+1, n-1, k), at(i+1, n-1, k+1), at(i, n-1, k+1)
            out.extend([a2, b2, c2, a2, c2, d2])

    for i in range(n - 1):
        for j in range(n - 1):
            a, b, c, d = at(i, j, 0), at(i+1, j, 0), at(i+1, j+1, 0), at(i, j+1, 0)
            out.extend([a, b, c, a, c, d])
            a2, b2, c2, d2 = at(i, j, n-1), at(i+1, j, n-1), at(i+1, j+1, n-1), at(i, j+1, n-1)
            out.extend([a2, c2, b2, a2, d2, c2])

    return out


def make_rng(seed: int) -> Callable[[], float]:
    """Seeded LCG random number generator (deterministic). Returns floats in [0, 1)."""
    state = [seed & 0xFFFFFFFF or 1]
    def _rng() -> float:
        state[0] = (state[0] * 1664525 + 1013904223) & 0xFFFFFFFF
        return state[0] / 0x100000000
    return _rng


def rand_int(rng: Callable[[], float], lo: int, hi: int) -> int:
    return int(rng() * (hi - lo + 1)) + lo


def shuffle(rng: Callable[[], float], arr: list) -> list:
    for i in range(len(arr) - 1, 0, -1):
        j = int(rng() * (i + 1))
        arr[i], arr[j] = arr[j], arr[i]
    return arr


def select_centers(grid_n: int, radius_cells: int, spacing_cells: int,
                   jitter: int, rng: Callable[[], float]) -> List[Tuple[int, int, int]]:
    """Option C: jittered sublattice + greedy dedup + forced coverage.

    1. Generate candidate centers on a regular sublattice (spacing = spacing_cells),
       each displaced by ±jitter cells (clamped to grid).
    2. Shuffle candidates (seeded) and greedily select, skipping any within
       Chebyshev distance spacing_cells of an already-selected center.
    3. Coverage check: every particle must be within radius_cells of some center.
       Uncovered particles are force-added as new centers (may violate spacing —
       coverage is mandatory, spacing is a soft constraint).

    With radius_cells=R and spacing_cells=S, adjacent centers share (2R+1−S) layers
    of particles. For 1–2 layers shared: S = 2R or S = 2R−1.
    """
    # 1. Sublattice candidates with jitter
    candidates: List[Tuple[int, int, int]] = []
    for k in range(0, grid_n, spacing_cells):
        for j in range(0, grid_n, spacing_cells):
            for i in range(0, grid_n, spacing_cells):
                ci = max(0, min(grid_n - 1, i + rand_int(rng, -jitter, jitter)))
                cj = max(0, min(grid_n - 1, j + rand_int(rng, -jitter, jitter)))
                ck = max(0, min(grid_n - 1, k + rand_int(rng, -jitter, jitter)))
                candidates.append((ci, cj, ck))

    # Ensure the far edge is represented even if gridN isn't a multiple of spacing.
    if (grid_n - 1) % spacing_cells != 0:
        for j in range(0, grid_n, spacing_cells):
            for i in range(0, grid_n, spacing_cells):
                candidates.append((
                    max(0, min(grid_n - 1, i + rand_int(rng, -jitter, jitter))),
                    max(0, min(grid_n - 1, j + rand_int(rng, -jitter, jitter))),
                    grid_n - 1,
                ))
        for i in range(0, grid_n, spacing_cells):
            for k in range(0, grid_n, spacing_cells):
                candidates.append((
                    max(0, min(grid_n - 1, i + rand_int(rng, -jitter, jitter))),
                    grid_n - 1,
                    max(0, min(grid_n - 1, k + rand_int(rng, -jitter, jitter))),
                ))
        for j in range(0, grid_n, spacing_cells):
            for k in range(0, grid_n, spacing_cells):
                candidates.append((
                    grid_n - 1,
                    max(0, min(grid_n - 1, j + rand_int(rng, -jitter, jitter))),
                    max(0, min(grid_n - 1, k + rand_int(rng, -jitter, jitter))),
                ))

    # 2. Shuffle + greedy select with min spacing
    shuffle(rng, candidates)
    centers: List[Tuple[int, int, int]] = []
    for (ci, cj, ck) in candidates:
        too_close = False
        for (si, sj, sk) in centers:
            if (abs(ci - si) <= spacing_cells and
                abs(cj - sj) <= spacing_cells and
                abs(ck - sk) <= spacing_cells):
                too_close = True
                break
        if not too_close:
            centers.append((ci, cj, ck))

    # 3. Coverage check: force-add centers for uncovered particles
    added = 0
    for k in range(grid_n):
        for j in range(grid_n):
            for i in range(grid_n):
                covered = False
                for (ci, cj, ck2) in centers:
                    if (abs(i - ci) <= radius_cells and
                        abs(j - cj) <= radius_cells and
                        abs(k - ck2) <= radius_cells):
                        covered = True
                        break
                if not covered:
                    centers.append((i, j, k))
                    added += 1
    if added > 0:
        print(f"info: coverage check added {added} extra center(s) for uncovered particles",
              file=sys.stderr)

    return centers


def generate(grid_n: int, cell_size: float, cluster_radius: int,
             cluster_spacing: int, jitter: int, seed: int,
             mass: float) -> Dict[str, Any]:
    particle_count = grid_n ** 3
    inv_mass = particle_count / mass
    half = (grid_n - 1) / 2.0

    # ── particle rest positions (axis-aligned, centered at origin) ───────
    positions: List[List[float]] = []
    for k in range(grid_n):
        for j in range(grid_n):
            for i in range(grid_n):
                x = (i - half) * cell_size
                y = (j - half) * cell_size
                z = (k - half) * cell_size
                positions.append([x, y, z, inv_mass])

    # ── surface indices ─────────────────────────────────────────────────
    surface_indices = build_surface_indices(grid_n)

    # ── select cluster centers (jittered sublattice + greedy + coverage) ─
    rng = make_rng(seed)
    centers = select_centers(grid_n, cluster_radius, cluster_spacing, jitter, rng)
    cluster_count = len(centers)

    # First pass: count members per cluster
    cluster_sizes = [0] * cluster_count
    for c, (ci, cj, ck) in enumerate(centers):
        cnt = 0
        for di in range(-cluster_radius, cluster_radius + 1):
            ni = ci + di
            if ni < 0 or ni >= grid_n:
                continue
            for dj in range(-cluster_radius, cluster_radius + 1):
                nj = cj + dj
                if nj < 0 or nj >= grid_n:
                    continue
                for dk in range(-cluster_radius, cluster_radius + 1):
                    nk = ck + dk
                    if nk < 0 or nk >= grid_n:
                        continue
                    cnt += 1
        cluster_sizes[c] = cnt

    # Prefix-sum for offsets
    offsets = [0] * cluster_count
    total = 0
    for c in range(cluster_count):
        offsets[c] = total
        total += cluster_sizes[c]

    cluster_entries = total

    # Second pass: write cluster data + indices + rest offsets
    clusters: List[Dict[str, Any]] = []
    cluster_indices: List[int] = [0] * cluster_entries
    rest_offsets: List[List[float]] = [[0.0, 0.0, 0.0] for _ in range(cluster_entries)]

    for c, (ci, cj, ck) in enumerate(centers):
        cnt = cluster_sizes[c]
        offset = offsets[c]

        # Rest center of mass
        com_x = com_y = com_z = 0.0
        for di in range(-cluster_radius, cluster_radius + 1):
            ni = ci + di
            if ni < 0 or ni >= grid_n:
                continue
            for dj in range(-cluster_radius, cluster_radius + 1):
                nj = cj + dj
                if nj < 0 or nj >= grid_n:
                    continue
                for dk in range(-cluster_radius, cluster_radius + 1):
                    nk = ck + dk
                    if nk < 0 or nk >= grid_n:
                        continue
                    pid = idx_at(ni, nj, nk, grid_n)
                    com_x += positions[pid][0]
                    com_y += positions[pid][1]
                    com_z += positions[pid][2]
        com_x /= cnt
        com_y /= cnt
        com_z /= cnt

        clusters.append({
            "restCom": [com_x, com_y, com_z],
            "count": cnt,
            "offset": offset,
        })

        local_idx = 0
        for di in range(-cluster_radius, cluster_radius + 1):
            ni = ci + di
            if ni < 0 or ni >= grid_n:
                continue
            for dj in range(-cluster_radius, cluster_radius + 1):
                nj = cj + dj
                if nj < 0 or nj >= grid_n:
                    continue
                for dk in range(-cluster_radius, cluster_radius + 1):
                    nk = ck + dk
                    if nk < 0 or nk >= grid_n:
                        continue
                    pid = idx_at(ni, nj, nk, grid_n)
                    slot = offset + local_idx
                    cluster_indices[slot] = pid
                    rest_offsets[slot] = [
                        positions[pid][0] - com_x,
                        positions[pid][1] - com_y,
                        positions[pid][2] - com_z,
                    ]
                    local_idx += 1

    return {
        "meta": {
            "gridN": grid_n,
            "cellSize": cell_size,
            "mass": mass,
            "seed": seed,
            "jitter": jitter,
            "clusterRadius": cluster_radius * cell_size,
            "clusterSpacing": cluster_spacing * cell_size,
            "clusterRadiusCells": cluster_radius,
            "clusterSpacingCells": cluster_spacing,
            "particleCount": particle_count,
            "clusterCount": cluster_count,
            "clusterEntries": cluster_entries,
            "surfaceVertexCount": len(surface_indices),
        },
        "positions": positions,
        "surfaceIndices": surface_indices,
        "clusters": clusters,
        "clusterIndices": cluster_indices,
        "restOffsets": rest_offsets,
    }


def main():
    parser = argparse.ArgumentParser(
        description="Generate a pre-baked soft-body asset file for demo9 PBD.",
    )
    parser.add_argument("--gridN", type=int, default=5, help="grid resolution (gridN^3 particles)")
    parser.add_argument("--cellSize", type=float, default=0.4, help="particle spacing / cell edge length (m)")
    parser.add_argument("--clusterRadius", type=float, default=0.8, help="Chebyshev membership radius (meters)")
    parser.add_argument("--clusterSpacing", type=float, default=None,
                        help="min Chebyshev distance between cluster centers (meters, default: = clusterRadius)")
    parser.add_argument("--jitter", type=int, default=1, help="sublattice jitter in cells (0 = regular grid)")
    parser.add_argument("--seed", type=int, default=42, help="RNG seed for reproducible jitter")
    parser.add_argument("--mass", type=float, default=1.0, help="total mass (kg)")
    parser.add_argument("--output", type=str, default="softbody_asset.json", help="output file path")
    args = parser.parse_args()

    if args.gridN < 2:
        print(f"error: --gridN must be >= 2, got {args.gridN}", file=sys.stderr)
        sys.exit(1)
    if args.cellSize <= 0:
        print(f"error: --cellSize must be > 0, got {args.cellSize}", file=sys.stderr)
        sys.exit(1)
    if args.clusterRadius <= 0:
        print(f"error: --clusterRadius must be > 0, got {args.clusterRadius}", file=sys.stderr)
        sys.exit(1)
    if args.jitter < 0:
        print(f"error: --jitter must be >= 0, got {args.jitter}", file=sys.stderr)
        sys.exit(1)

    cluster_spacing_m = args.clusterSpacing if args.clusterSpacing is not None else args.clusterRadius

    # Convert meter values to grid cells (quantized — actual physical size may differ).
    cluster_radius = max(1, round(args.clusterRadius / args.cellSize))
    cluster_spacing = max(1, round(cluster_spacing_m / args.cellSize))

    if cluster_radius > args.gridN - 1:
        print(f"error: --clusterRadius {args.clusterRadius}m = {cluster_radius} cells, "
              f"must be < gridN ({args.gridN}) = {(args.gridN-1)*args.cellSize}m", file=sys.stderr)
        sys.exit(1)

    # Warn if quantization changed the actual value.
    if cluster_radius * args.cellSize != args.clusterRadius:
        print(f"warning: --clusterRadius {args.clusterRadius}m quantized to {cluster_radius} cells "
              f"= {cluster_radius * args.cellSize:.3f}m", file=sys.stderr)
    if cluster_spacing * args.cellSize != cluster_spacing_m:
        print(f"warning: --clusterSpacing {cluster_spacing_m}m quantized to {cluster_spacing} cells "
              f"= {cluster_spacing * args.cellSize:.3f}m", file=sys.stderr)

    data = generate(args.gridN, args.cellSize, cluster_radius, cluster_spacing,
                    args.jitter, args.seed, args.mass)

    with open(args.output, "w", encoding="utf-8") as f:
        json.dump(data, f, indent=2)

    m = data["meta"]
    cube_edge = (m["gridN"] - 1) * m["cellSize"]
    max_cluster = (2 * m["clusterRadiusCells"] + 1) ** 3
    shared_layers = max(0, 2 * m["clusterRadiusCells"] + 1 - m["clusterSpacingCells"])
    print(
        f"Generated {args.output}:\n"
        f"  gridN={m['gridN']}  cellSize={m['cellSize']}m  cubeEdge={cube_edge:.3f}m  cubeVolume={cube_edge**3:.3f}m³\n"
        f"  clusterRadius={m['clusterRadius']}m ({m['clusterRadiusCells']} cells)"
        f"  clusterSpacing={m['clusterSpacing']}m ({m['clusterSpacingCells']} cells)"
        f"  jitter={m['jitter']}  seed={m['seed']}  mass={m['mass']}\n"
        f"  particles={m['particleCount']}  clusters={m['clusterCount']}  clusterEntries={m['clusterEntries']}\n"
        f"  maxClusterMembers={max_cluster}  sharedLayers≈{shared_layers}"
        f"  surfaceTriangles={m['surfaceVertexCount'] // 3}\n"
        f"  perParticleMass={m['mass'] / m['particleCount']:.6f} kg  invMass={m['particleCount'] / m['mass']:.1f}"
    )


if __name__ == "__main__":
    main()
